File: frontend/android/app/src/main/java/com/kaixuan/opencode/pocket/plugins/AppSettingsPlugin.java
package com.kaixuan.opencode.pocket.plugins;

import android.Manifest;
import android.os.Build;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

/**
 * AppSettings — 打开对应系统权限页（定位本应用），并检查/申请麦克风、通知、相机、相册。
 *
 * 第一次拒绝仍弹系统窗；永久禁止则 openAppDetails({ name }) 打开该权限的系统设置。
 */
@CapacitorPlugin(
        name = "AppSettings",
        permissions = {
            @Permission(alias = "microphone", strings = {Manifest.permission.RECORD_AUDIO}),
            @Permission(alias = "notifications", strings = {Manifest.permission.POST_NOTIFICATIONS}),
            @Permission(alias = "camera", strings = {Manifest.permission.CAMERA}),
            @Permission(alias = "photos", strings = {Manifest.permission.READ_MEDIA_IMAGES}),
            @Permission(alias = "photosLegacy", strings = {Manifest.permission.READ_EXTERNAL_STORAGE})
        })
public class AppSettingsPlugin extends Plugin {
    @PluginMethod
    public void openAppDetails(PluginCall call) {
        String name = call.getString("name", "");
        boolean opened = PermissionSettingsLauncher.open(getContext(), getActivity(), name);
        JSObject ret = new JSObject();
        ret.put("opened", opened);
        call.resolve(ret);
    }

    @PluginMethod
    public void check(PluginCall call) {
        String name = aliasOf(call);
        if (name == null) {
            call.reject("name must be microphone, notifications, camera, or photos");
            return;
        }
        call.resolve(statePayload(name));
    }

    @PluginMethod
    public void request(PluginCall call) {
        String name = aliasOf(call);
        if (name == null) {
            call.reject("name must be microphone, notifications, camera, or photos");
            return;
        }
        if ("notifications".equals(name) && Build.VERSION.SDK_INT < 33) {
            call.resolve(grantedPayload("notifications"));
            return;
        }
        requestPermissionForAlias(resolvedAlias(name), call, "completeRequest");
    }

    @PermissionCallback
    private void completeRequest(PluginCall call) {
        String name = aliasOf(call);
        if (name == null) {
            call.reject("name must be microphone, notifications, camera, or photos");
            return;
        }
        call.resolve(statePayload(name));
    }

    private static String aliasOf(PluginCall call) {
        String name = call.getString("name", "");
        if ("microphone".equals(name)
                || "notifications".equals(name)
                || "camera".equals(name)
                || "photos".equals(name)) {
            return name;
        }
        return null;
    }

    private static String resolvedAlias(String name) {
        if ("photos".equals(name)) {
            return Build.VERSION.SDK_INT >= 33 ? "photos" : "photosLegacy";
        }
        return name;
    }

    private JSObject statePayload(String alias) {
        if ("notifications".equals(alias) && Build.VERSION.SDK_INT < 33) {
            return grantedPayload(alias);
        }
        PermissionState state = getPermissionState(resolvedAlias(alias));
        if (state == null) {
            state = PermissionState.PROMPT;
        }
        JSObject ret = new JSObject();
        ret.put("name", alias);
        ret.put("status", state.toString());
        boolean canRequestAgain =
                state == PermissionState.PROMPT || state == PermissionState.PROMPT_WITH_RATIONALE;
        ret.put("canRequestAgain", canRequestAgain);
        return ret;
    }

    private static JSObject grantedPayload(String alias) {
        JSObject ret = new JSObject();
        ret.put("name", alias);
        ret.put("status", PermissionState.GRANTED.toString());
        ret.put("canRequestAgain", false);
        return ret;
    }
}

File: frontend/ios/App/App/Plugins/AppSettingsPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 * AppSettingsPlugin —— iOS 镜像 (Phase 7)。
 *
 * 对应 Android `com.kaixuan.opencode.pocket.plugins.AppSettingsPlugin`。
 * 范围：
 *   - openAppDetails({ name }) → 打开 iOS 设置 App 到本应用权限页
 *   - check({ name }) → 检查权限状态
 *   - request({ name }) → 申请权限（iOS 14+ 部分权限需用户主动设置）
 *
 * iOS 与 Android 差异：
 *   - iOS 没有「永久拒绝」（Android 才有），权限状态只是 granted/denied/prompt。
 *   - iOS 的 photos 权限（NSPhotoLibraryUsageDescription）首次会弹系统；
 *     microphone / camera 首次会弹系统。
 *   - 「打开应用设置页」iOS 通过 `UIApplication.openSettingsURLString`。
 */
@objc(AppSettingsPlugin)
public class AppSettingsPlugin: CAPPlugin {
    private static let TAG = "AppSettingsPlugin"

    @objc func openAppDetails(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("name is required")
            return
        }
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString) else {
                call.resolve(["opened": false])
                return
            }
            if UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.open(url, options: [:]) { success in
                    call.resolve(["opened": success])
                }
            } else {
                call.resolve(["opened": false])
            }
        }
    }

    @objc func check(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("name is required")
            return
        }
        let state = self.permissionState(name: name)
        call.resolve(["state": state])
    }

    @objc func request(_ call: CAPPluginCall) {
        // iOS 上权限请求通常在用户首次使用相应能力时由系统弹窗；
        // 这里只是占位（真正权限弹窗在 Camera.getPhoto / Recorder.start 等
        // 系统 API 触发时由 iOS 自动弹出）。
        call.resolve(["granted": true])
    }

    private func permissionState(name: String) -> String {
        // 简化：iOS 没有统一的权限查询 API（不像 Android 的 ContextCompat.checkSelfPermission）。
        // 真实使用时应通过具体能力 API（AVCaptureDevice.authorizationStatus 等）查询。
        // 这里返回 'prompt' 作为兜底，调用方根据行为判断。
        return "prompt"
    }
}